registrations/api.py

```diff
@@ -277,36 +282,28 @@ def filter_queryset(self, queryset):
     def filter_registration(self, queryset, name, value: list[str]):
         user = self.request.user
 
-        # Get admin organizations and descendants only once instead of using is_admin method
-        registration_admin_organizations = (
-            user.get_registration_admin_organizations_and_descendants()
-        )
+        try:
+            registrations = Registration.objects.filter(pk__in=value)
+        except ValueError:
+            registrations = Registration.objects.none()
 
-        registrations = []
-        for pk in value:
-            try:
-                registration = Registration.objects.select_related(
-                    "event__publisher"
-                ).get(pk=pk)
-            except (ValueError, Registration.DoesNotExist):
-                raise ParseError(
-                    _("Registration with id {pk} doesn't exist.").format(pk=pk)
-                )
+        if not registrations:
+            raise ParseError(_("No registrations found with the given IDs."))
 
-            if not (
-                user.is_superuser
-                or user.is_registration_user_access_user_of(
-                    registration.registration_user_accesses
-                )
-                or registration.publisher in registration_admin_organizations
-            ):
-                raise DRFPermissionDenied(
-                    _(
-                        "Only the admins of the organization {organization} have access rights."
-                    ).format(organization=registration.publisher)
-                )
+        if user.is_superuser:
+            return queryset.filter(**{f"{name}__in": registrations})
+
+        qs_filter = Q(event__publisher__in=self._user_registration_admin_organizations)
+        if user.is_strongly_identified:
+            qs_filter |= Q(registration_user_accesses__email=user.email)
+        registrations = registrations.filter(qs_filter)
 
-            registrations.append(registration)
+        if not registrations:
+            raise DRFPermissionDenied(
+                _(
+                    "Only the admins of the registration organizations have access rights."
+                )
+            )
```

Comment:
It's unfortunate that this fits so well right here where it is. On a fundamental level I maintain a position that permission checks should be performed before the view in permission class(es). Just food for thought, not requesting this to be changed as it's probably non-trivial. Plus LE api has a tradition of raising PermissionDenied all over the place :laughing:

Reply:
Agree with both non-triviality and that the perms should be checked in permission classes.